src/service/xhr-impl.js

```diff
@@ -407,13 +427,11 @@ function isRetriable(status) {
  */
 export function assertSuccess(response) {
   return new Promise((resolve, reject) => {
-    if (response.status < 200 || response.status >= 300) {
-      /** @const {!Error} */
-      const err = user().createError(`HTTP error ${response.status}`);
-      err.response = response;
-      if (isRetriable(response.status)) {
-        err.retriable = true;
-      }
+    const status = response.status;
+    if (status < 200 || status >= 300) {
+      const retriable = isRetriable(status);
+      const err = user().createCustomError(
+          new FetchError(`HTTP error ${status}`, response, retriable));
```

Comment:
I was trying to adapt the existing code to work better with the type system, since `gulp check-types` would complain that I was accessing properties not available on an `Error` instance. The previous implementation instantiated a `new Error` and augmented it with the `response` and the `responseJson`, in effect rejecting an error and passing data together. It was added in #3669. This felt hacky and unclear since the type was sort of lying (it says it's an Error, but it has these extra fields) so I wanted to make the augmentation more formal.